### ISSUE TITLE
merging batch requests only updates roi

### ISSUE THREAD

**Issue by neptunes5thmoon** [funkey/gunpowder#144]
I'm trying to write a node that requires its input to be of a specific datatype, plus a cast node to meet that requirement but have been running into issues.
I think what's happening is that the current process for merging batch requests (for spatial arrays that are updated in place) only updates the roi, not the other attributes in a spec. As an example, in `Normalize` setting `dtype = None` has no effect which can lead to errors if downstream nodes request a specific datatype. https://github.com/funkey/gunpowder/blob/b074e502cd90e27cdc14ed54cd8dace3a6b2b530/gunpowder/nodes/normalize.py#L44-L48
Is this the intended behavior and requesting a specific datatype is a bad idea?
I identified this as the relevant part of the code:
https://github.com/funkey/gunpowder/blob/b074e502cd90e27cdc14ed54cd8dace3a6b2b530/gunpowder/batch_request.py#L100-L116

**Comment by pattonw** [comment]
I think best practice would be to create a new array with the new datatype and its own key to avoid needing to merge requests and arrays.

I'm not sure if we want to support a dynamic pipeline that provides an array with ambiguous dtype from which you can then request float32 or uint64 or any other dtype and expect it to be cast properly. At the moment I wouldn't expect this to work easily. A cast node that casts to a specific dtype though should work fine (like normalize).

Do you need the pipeline to support a dynamic dtype defined by your request?

**Comment by neptunes5thmoon** [comment]
I don't need a dynamic dtype that is cast automatically, sorry for not being more clear.

What I have instead is a node that requires a specific dtype. Here's a pretty minimal example:

```python
import gunpowder as gp
import numpy as np
import urllib.request
import os


class NodeNeedsFloat32(gp.BatchFilter):
    def __init__(self, array_in, array_out):
        self.array_in = array_in
        self.array_out = array_out

    def setup(self):
        self.provides(self.array_out, self.spec[self.array_in].copy())

    def prepare(self, request):
        deps = gp.BatchRequest()
        deps[self.array_in] = request[self.array_out].copy()
        deps[self.array_in].dtype = np.float32 # removing this line makes it work
        return deps

    def process(self, batch, request):
        # some function that requires array_in to be float32
        assert batch[self.array_in].spec.dtype == np.float32
        data = batch[self.array_in].data
        spec = batch[self.array_in].spec.copy()
        spec.roi = request[self.array_out].roi.copy()
        batch = gp.Batch()
        batch[self.array_out] = gp.Array(data, spec)
        return batch


url = "https://cremi.org/static/data/sample_A_20160501.hdf"
urllib.request.urlretrieve(url, os.path.join('.', 'sample_A.hdf'))

raw = gp.ArrayKey("RAW")
raw_float_copy = gp.ArrayKey("RAW_FLOAT_COPY")
source = gp.Hdf5Source(
    os.path.join('.', 'sample_A.hdf'),
    datasets={
        raw: "volumes/raw"  # this is uint8
    }
)
normalize = gp.Normalize(raw)  # this converts to float32

floatnode = NodeNeedsFloat32(raw, raw_float_copy)

pipeline = source + normalize + floatnode

request = gp.BatchRequest()
request[raw_float_copy] = gp.Roi(gp.Coordinate((0, 0, 0)), gp.Coordinate((100, 100, 100))*gp.Coordinate((40, 4, 4)))

with gp.build(pipeline):
    pipeline.request_batch(request)
```

The reason this doesn't work is that the only part of the dependencies that are returned by `prepare` that are being considered in this case is the roi. This happens in the second code snippet I posted above.
I haven't come across a node that requests a specific dtype, so I was wondering whether that is beyond the scope of what requests are there for. I haven't explicitly tested this but I think the same would happen if a node was explicit about the required `voxel_size`, `interpolatable` etc.

**Comment by pattonw** [comment]
Oh I see, so you want to be able to define the meta-data that you expect in the `prepare` step, and have your pipeline immediately throw a helpful error if you place a `NodeNeedsFloat32` into a pipeline that does not provide float32 for the requested array?
Yes, I think this is in scope for requests and should be done exactly as you have it here.
We do already check requests for consistency with provided specs in `BatchProvider`, so I think it would be a great idea to add the same check to the dependencies returned from a `prepare` call. We could also add support for the meta-data fields in merge.

**Comment by neptunes5thmoon** [comment]
exactly! I'm happy to get started with a PR (prob in the new year) but might need some help with the inner workings of gunpowder

**Comment by pattonw** [comment]
Actually, it @abred already solved this in https://github.com/funkey/gunpowder/pull/119. I went ahead and merged it into the `v1.2-dev` branch so you should see the expected behavior if you use that branch.
